ext/nnstreamer/tensor_filter/tensor_filter_edgetpu.cc

```diff
@@ -0,0 +1,400 @@
+/**
+ * GStreamer Tensor_Filter, Edge-TPU Module
+ * Copyright (C) 2019 MyungJoo Ham <myungjoo.ham@samsung.com>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Library General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Library General Public License for more details.
+ *
+ */
+/**
+ * @file	tensor_filter_edgetpu.c
+ * @date	10 Dec 2019
+ * @brief	Edge-TPU module for tensor_filter gstreamer plugin
+ * @see		http://github.com/nnsuite/nnstreamer
+ * @author	MyungJoo Ham <myungjoo.ham@samsung.com>
+ * @bug		No known bugs except for NYI items
+ *
+ * This is the per-NN-framework plugin (Edge TPU) for tensor_filter.
+ *
+ * @todo This supports single-model-single-TPU only. Prep for multi-TPU.
+ * @todo A lot of this duplicate tf-lite filter.
+ *       We may be able to embed this code into tf-lite filter code.
+ */
+#include <iostream>
+
+#include <stdint.h>
+
+#include <nnstreamer_plugin_api_filter.h>
+#include <tensor_common.h>
+#include <glib.h>
+
+#include <edgetpu.h>
+#include <tensorflow/lite/model.h>
+#include <tensorflow/lite/interpreter.h>
+#include <tensorflow/lite/builtin_op_data.h>
+#include <tensorflow/lite/kernels/register.h>
+
+void init_filter_edgetpu (void) __attribute__ ((constructor));
+void fini_filter_edgetpu (void) __attribute__ ((destructor));
+
+/**
+ * @brief Internal data structure
+ */
+typedef struct {
+  char *model_path; /**< The model *.tflite file */
+  GstTensorsInfo input; /**< Input tensors metadata */
+  GstTensorsInfo output; /**< Output tensors metadata */
+
+  /* EdgeTPU + Tensorflow-lite Execution */
+  std::unique_ptr<tflite::Interpreter> model_interpreter;
+      /**< TFLite interpreter */
+  tflite::Interpreter *interpreter;
+      /**< model_interpreter.get() */
+  std::shared_ptr<edgetpu::EdgeTpuContext> edgetpu_context;
+      /**< EdgeTPU Device context */
+  std::unique_ptr<tflite::FlatBufferModel> model;
+      /**< Loaded TF Lite model (from model_path) */
+} pdata;
+
+/**
+ * @brief Get TF-Lite interpreter w/ edgetpu context
+ */
+static std::unique_ptr<tflite::Interpreter>
+BuildEdgeTpuInterpreter(const tflite::FlatBufferModel &model,
+      edgetpu::EdgeTpuContext* edgetpu_context)
+{
+  tflite::ops::builtin::BuiltinOpResolver resolver;
+  resolver.AddCustom(edgetpu::kCustomOp, edgetpu::RegisterCustomOp());
+  std::unique_ptr<tflite::Interpreter> interpreter;
+  if (tflite::InterpreterBuilder(model, resolver)(&interpreter) !=
+      kTfLiteOk) {
+    std::cerr << "Failed to build interpreter." << std::endl;
+  }
+
+  interpreter->SetExternalContext(kTfLiteEdgeTpuContext, edgetpu_context);
+  interpreter->SetNumThreads(1);
```

Comment:
can number of threads be made a constant variable (which can be configured later if needed)?

Reply:
With edge-TPU, the workloads are supposed to be contained in the edge-TPU. not CPU.